sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java

```diff
@@ -0,0 +1,324 @@
+   *             Timestamp} representing the timestamp at which the new partition started being
+   *             valid in Cloud Spanner.
+   *         <li>{@link ChangeStreamRecordMapper#RECORD_SEQUENCE_COLUMN}: non-nullable {@link
+   *             String} representing the order in which this record appears within the context of a
+   *             partition and commit timestamp.
+   *         <li>{@link ChangeStreamRecordMapper#CHILD_PARTITIONS_COLUMN}: non-nullable {@link List}
+   *             of {@link Struct} representing the new child partitions.
+   *             <ul>
+   *               <li>{@link ChangeStreamRecordMapper#TOKEN_COLUMN}: non-nullable {@link String}
+   *                   representing the unique identifier of the new child partition.
+   *               <li>{@link ChangeStreamRecordMapper#PARENT_PARTITION_TOKENS_COLUMN}: non-nullable
+   *                   {@link List} of {@link String} representing the unique identifier(s) of
+   *                   parent partition(s) where this child partition originated from.
+   *             </ul>
+   *       </ul>
+   * </ul>
+   *
+   * @param partition the partition metadata from which the row was generated
+   * @param row the struct row, representing a single change stream result (it may contain multiple
+   *     change stream records within)
+   * @param resultSetMetadata the metadata generated when reading the change stream row
+   * @return a {@link List} of {@link ChangeStreamRecord} subclasses
+   */
+  public List<ChangeStreamRecord> toChangeStreamRecords(
+      PartitionMetadata partition, Struct row, ChangeStreamResultSetMetadata resultSetMetadata) {
+    return row.getStructList(0).stream()
+        .flatMap(struct -> toChangeStreamRecord(partition, struct, resultSetMetadata))
+        .collect(Collectors.toList());
+  }
+
+  private Stream<ChangeStreamRecord> toChangeStreamRecord(
+      PartitionMetadata partition, Struct row, ChangeStreamResultSetMetadata resultSetMetadata) {
+
+    final Stream<DataChangeRecord> dataChangeRecords =
+        row.getStructList(DATA_CHANGE_RECORD_COLUMN).stream()
+            .filter(this::isNonNullDataChangeRecord)
+            .map(struct -> toDataChangeRecord(partition, struct, resultSetMetadata));
+
+    final Stream<HeartbeatRecord> heartbeatRecords =
+        row.getStructList(HEARTBEAT_RECORD_COLUMN).stream()
+            .filter(this::isNonNullHeartbeatRecord)
+            .map(struct -> toHeartbeatRecord(partition, struct, resultSetMetadata));
+
+    final Stream<ChildPartitionsRecord> childPartitionsRecords =
+        row.getStructList(CHILD_PARTITIONS_RECORD_COLUMN).stream()
+            .filter(this::isNonNullChildPartitionsRecord)
+            .map(struct -> toChildPartitionsRecord(partition, struct, resultSetMetadata));
+
+    return Stream.concat(
+        Stream.concat(dataChangeRecords, heartbeatRecords), childPartitionsRecords);
```

Comment:
I'm sure you've thought about this, but does it not matter that we're reordering the records here? (moving all Data Change Records  to the begining, then heartbeat, then child partitions)

Reply:
That is a good point, I think this might affect the watermark. I will change this.

Thanks for pointing this out.

Reply:
Oh this is not a problem, because `DataChangeRecord`s timestamps will be < `HeartbeatRecord`s timestamps will be < `ChildPartitionRecord`s timestamps.

Other than that, I don't think we have a mix of these types of records in a single response at the moment.